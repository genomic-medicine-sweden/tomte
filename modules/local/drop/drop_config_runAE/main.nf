process DROP_CONFIG_RUN_AE {
    tag "DROP_CONFIG_RUN_AE"
    label 'process_drop'

    // Exit if running this module with -profile conda / -profile mamba
    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
        exit 1, "Local DROP module does not support Conda. Please use Docker / Singularity / Podman instead."
    }

    container "docker.io/clinicalgenomics/drop:1.4.0"

    input:
    tuple val(meta), path(fasta), path(fai)
    tuple val(meta2), path(gtf)
    path sample_annotation
    tuple path(bam), path(bai)
    path ref_drop_count_file
    path ref_splice_folder
    val(genome)
    val(drop_group_samples_ae)
    val(drop_group_samples_as)
    val(drop_padjcutoff_ae)
    val(drop_zScoreCutoff)
    val(skip_export_counts_drop)

    output:
    path('config.yaml')             , emit: config_drop
    path('output')                  , emit: drop_ae_out
    path('OUTRIDER_results_all.Rds'), emit: drop_ae_rds
    path('gene_name_mapping*')      , emit: drop_gene_name
    path('exported_counts_ae')      , emit: gene_counts_ae, optional: true
    path "versions.yml"             , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def genome_assembly = "${genome}".contains("h37") ? "hg19" : "${genome}"
    def drop_group = "${drop_group_samples_ae}".replace(" ","")
    def drop_other_group_samples = "${drop_group_samples_as}".replace(" ","")
    def zscorecutoff = drop_zScoreCutoff ? "--zscorecutoff ${drop_zScoreCutoff}" : ''

    """
    TMPDIR=\$PWD
    HOME=\$PWD

    drop init

    drop_config.py \\
        --genome_fasta ${fasta} \\
        --gtf ${gtf} \\
        --drop_module AE \\
        --genome_assembly $genome_assembly \\
        --drop_group_samples $drop_group \\
        --drop_other_group_samples $drop_other_group_samples \\
        --padjcutoff ${drop_padjcutoff_ae} \\
        $zscorecutoff \\
        --output config.yaml

    snakemake aberrantExpression --cores ${task.cpus} --rerun-triggers mtime $args

    if [[ $skip_export_counts_drop == false ]]; then
        snakemake exportCounts --cores 1
        mkdir -p exported_counts_ae
        cp sample_annotation.tsv exported_counts_ae/.
        cp output/processed_results/exported_counts/*/geneCounts.tsv.gz exported_counts_ae/.
    fi

    cp output/processed_results/aberrant_expression/*/outrider/outrider/OUTRIDER_results_all.Rds .
    cp output/processed_data/preprocess/*/gene_name_mapping_*.tsv .

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        drop_config: \$(drop_config.py --version)
        drop: \$(echo \$(drop --version) |  sed -n 's/drop, version //p')
    END_VERSIONS
    """

    stub:
    """
    touch config.yaml
    touch OUTRIDER_results_all.Rds
    touch gene_name_mapping_.tsv
    mkdir output
    if [[ $skip_export_counts_drop == false ]]; then
        mkdir -p exported_counts_ae
        touch exported_counts_ae/sample_annotation.tsv
        touch exported_counts_ae/geneCounts.tsv.gz
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        drop_config: \$(drop_config.py --version)
        drop: \$(echo \$(drop --version) |  sed -n 's/drop, version //p')
    END_VERSIONS
    """
}
