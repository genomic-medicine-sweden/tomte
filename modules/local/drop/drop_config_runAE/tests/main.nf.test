nextflow_process {

    name "Test Process DROP_CONFIG_RUN_AE"
    script "../main.nf"
    process "DROP_CONFIG_RUN_AE"
    config "./nextflow.config"
    tag "modules"
    tag "modules_local"
    tag "drop"
    tag "drop_config_runAE"

    setup {
        run("GUNZIP", alias: "GUNZIP_FASTA") {
            script "../../../../../modules/nf-core/gunzip/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id:'fasta' ],
                    file("${projectDir}/test_data/grch37_chr21.fa.gz", checkIfExists: true)
                ])
                """
            }
        }

        run("SAMTOOLS_FAIDX") {
            script "../../../../../modules/nf-core/samtools/faidx/main.nf"
            process {
                """
                input[0] = GUNZIP_FASTA.out.gunzip
                input[1] = [[],[]]
                input[2] = []
                """
            }
        }

        run("GUNZIP", alias: "GUNZIP_GTF") {
            script "../../../../../modules/nf-core/gunzip/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id:'gtf' ],
                    file("${projectDir}/test_data/grch37_chr21.gtf.gz", checkIfExists: true)
                ])
                """
            }
        }

        run("DROP_SAMPLE_ANNOT") {
            script "../../../../../modules/local/drop/drop_sample_annot/main.nf"
            process {
                """
                input[0] = Channel.of([
                            ['ACC5963A1'], [false], ['reverse'], ['NA'], ["${projectDir}/test_data/test_ACC5963A1.vcf.gz"], ["${projectDir}/test_data/test_ACC5963A1.vcf.gz.tbi"], ['ACC5963A1'],
                            file("${projectDir}/test_data/ACC5963A1.bam", checkIfExists: true),
                            file("${projectDir}/test_data/ACC5963A1.bam.bai", checkIfExists: true)
                        ])
                input[1] = channel.fromPath("${projectDir}/test_data/drop_data/geneCounts.tsv.gz", checkIfExists: true)
                input[2] = channel.fromPath("${projectDir}/test_data/drop_data/sampleAnnotation.tsv", checkIfExists: true)
                input[3] = 'outrider'
                input[4] = 'fraser'
                """
            }
        }

    }

    test("Should run without failures") {

        when {
            process {
                """
                input[0] = GUNZIP_FASTA.out.gunzip.join(SAMTOOLS_FAIDX.out.fai).collect()
                input[1] = GUNZIP_GTF.out.gunzip.collect()
                input[2] = DROP_SAMPLE_ANNOT.out.drop_annot
                input[3] = Channel.of([
                        file("${projectDir}/test_data/ACC5963A1.bam", checkIfExists: true),
                        file("${projectDir}/test_data/ACC5963A1.bam.bai", checkIfExists: true)
                    ])
                input[4] = channel.fromPath("${projectDir}/test_data/drop_data/geneCounts.tsv.gz", checkIfExists: true)
                input[5] = channel.fromPath("${projectDir}/test_data/drop_data", checkIfExists: true)
                input[6] = 'GRCh37'
                input[7] = 'outrider'
                input[8] = 'fraser'
                input[9] = 1
                input[10] = 2.5
                input[11] = true
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.config_drop.collect{ file(it).name },
                    process.out.drop_gene_name,
                    process.out.drop_ae_rds.collect{ file(it).name },
                    process.out.drop_ae_out.collect{ file(it).name },
                    process.out.versions,
                ).match()}
            )
        }

    }

}
