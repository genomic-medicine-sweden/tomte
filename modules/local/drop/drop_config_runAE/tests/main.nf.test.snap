{
    "Should run without failures": {
        "content": [
            [
                "config.yaml"
            ],
            [
                "gene_name_mapping_grch37_chr21.tsv:md5,b44ef48c094b206faee684d7961b85ca"
            ],
            [
                "OUTRIDER_results_all.Rds"
            ],
            [
                "output"
            ],
            [
                "versions.yml:md5,49e1d7b9f8f2f9ae82fab6d3aee2d456"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.5"
        },
        "timestamp": "2025-09-10T14:54:46.384685723"
    }
}